File: app/src/main/java/com/example/miodragmilosevic/nativetest/filters/javaFilters/PixelateFilter.java
package com.example.miodragmilosevic.nativetest.filters.javaFilters;

import android.graphics.Color;

import com.example.miodragmilosevic.nativetest.filters.ImageFilter;

/**
 * Created by miodrag.milosevic on 12/15/2017.
 */

public class PixelateFilter extends ImageFilter {

    private static final int PIXEL_SIZE = 10;

    @Override
    protected int[] transformPixels(int[] pixelArray, int imageWidth, int imageHeight) {
        int index = 0;
        // iteration through pixels
        for (int i = 0; i < imageWidth; i += PIXEL_SIZE) {
            for (int j = 0; j < imageHeight; j += PIXEL_SIZE) {
                int a = 0, r = 0, g = 0, b = 0;
                int sum = PIXEL_SIZE * PIXEL_SIZE;
                for (int x = i; x < i + PIXEL_SIZE; x++) {


                    for (int y = j; y < j + PIXEL_SIZE; y++) {
                        index = x + y * imageWidth;
                        if (index < imageWidth * imageHeight) {
                            a = Color.alpha(pixelArray[index]);
                            r += Color.red(pixelArray[index]);
                            g += Color.green(pixelArray[index]);
                            b += Color.blue(pixelArray[index]);
                        }
                    }
                }
                for (int x = i; x < i + PIXEL_SIZE; x++) {
                    for (int y = j; y < j + PIXEL_SIZE; y++) {
                        index = x + y * imageWidth;
                        if (x < imageWidth && y < imageHeight && index < imageWidth * imageHeight) {
                            pixelArray[x + y * imageWidth] = Color.argb(a,r/sum, g/sum, b/sum);
                        }

                    }
                }
            }
        }
        return pixelArray;
    }
}

File: app/src/main/cpp/PixelateFilter.cpp
//
// Created by miodrag.milosevic on 12/29/2017.
//

#include "PixelateFilter.h"
#include "color.h"

void PixelateFilter::transformPixels(jint *pixelArray, jint width, jint height) {
    jint index = 0;
    jint pixelSize = 10;
// iteration through pixels
    for (jint i = 0; i < width; i += pixelSize) {
        for (jint j = 0; j < height; j += pixelSize) {
            jint a = 0, r = 0, g = 0, b = 0;
            jint sum = pixelSize * pixelSize;
            for (jint x = i; x < i + pixelSize; x++) {


                for (jint y = j; y < j + pixelSize; y++) {
                    index = x + y * width;
                    if (index < width * height) {
                        a = Color::alpha(pixelArray[index]);
                        r += Color::red(pixelArray[index]);
                        g += Color::green(pixelArray[index]);
                        b += Color::blue(pixelArray[index]);
                    }
                }
            }
            for (jint x = i; x < i + pixelSize; x++) {
                for (jint y = j; y < j + pixelSize; y++) {
                    index = x + y * width;
                    if (x < width && y < height && index < width * height) {
                        pixelArray[x + y * width] = Color::argb(a,r/sum, g/sum, b/sum);
                    }

                }
            }
        }
    }
}
